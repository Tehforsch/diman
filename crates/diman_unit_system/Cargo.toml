[package]
name = "diman_unit_system"
version = "0.2.0"
edition = "2021"
authors = [
    "Toni Peter <mail@tonipeter.de>",
]

[features]
glam = []
glam-vec2 = []
glam-dvec2 = []
glam-vec3 = []
glam-dvec3 = []
f32 = []
f64 = []
mpi = []
serde = []
rand = []
hdf5 = []
gen-vec-names = [] 

[dependencies]
derive_verify = { path = "../../crates/derive_verify" }
syn = { version = "1.0", features = ["full", "extra-traits"] }
quote = "1.0"
proc-macro2 = "1.0"

[lib]
proc-macro = true
