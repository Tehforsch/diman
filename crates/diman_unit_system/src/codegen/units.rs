use proc_macro2::TokenStream;
use quote::{format_ident, quote, quote_spanned};
use syn::Type;

use super::storage_types::{FloatType, VectorType};
use crate::types::{Defs, Unit};

impl Defs {
    pub fn gen_unit_constructors(&self) -> TokenStream {
        self.units
            .iter()
            .map(|unit| {
                let dimension = self.get_dimension_expr(&unit.dimensions);
                let vector_impls: TokenStream = self
                    .vector_types()
                    .iter()
                    .map(|vector_type| self.vector_unit_constructor(vector_type, unit, &dimension))
                    .collect();
                let float_impls: TokenStream = self
                    .float_types()
                    .iter()
                    .map(|float_type| self.float_unit_constructor(float_type, unit, &dimension))
                    .collect();
                let conversion_impls: TokenStream = self
                    .storage_types()
                    .map(|storage_type| {
                        let type_ = storage_type.name();
                        let float_type = storage_type.base_storage();
                        self.conversion_impls(type_, float_type, unit, &dimension)
                    })
                    .collect();
                quote! {
                    #conversion_impls
                    #float_impls
                    #vector_impls
                }
            })
            .collect()
    }

    fn conversion_impls(
        &self,
        storage_type: &Type,
        base_type: &FloatType,
        unit: &Unit,
        dimension: &TokenStream,
    ) -> TokenStream {
        let quantity_type = &self.quantity_type;
        let unit_name = &unit.name;
        let conversion_method_name = format_ident!("in_{}", unit_name);
        let magnitude = unit.magnitude;
        let base_type = &base_type.name;
        quote! {
            impl #quantity_type<#storage_type, {#dimension}> {
                pub fn #conversion_method_name(self) -> #storage_type {
                    let magnitude: #base_type = #magnitude as #base_type;
                    self.0 / magnitude
                }
            }
        }
    }

    fn float_unit_constructor(
        &self,
        float_type: &FloatType,
        unit: &Unit,
        quantity_dimension: &TokenStream,
    ) -> TokenStream {
        let Defs { quantity_type, .. } = &self;
        let Unit {
            name: unit_name,
            magnitude,
            ..
        } = unit;
        let name = &float_type.name;
        let span = self.dimension_type.span();
        // Without const_fn_floating_point_arithmetic (https://github.com/rust-lang/rust/issues/57241)
        // we cannot make unit constructors a const fn in general (since it requires the unstable
        // const_fn_floating_point_arithmetic feature). The following allows the constructor with 1.0
        // conversion factor to be const.
        let const_fn = *magnitude == 1.0;
        let fn_def = if const_fn {
            quote! { const fn }
        } else {
            quote! { fn }
        };
        let value = if const_fn {
            quote! { val }
        } else {
            quote! { val * #magnitude as #name }
        };
        quote_spanned! {span =>
            impl #quantity_type<#name, {#quantity_dimension}> {
                pub #fn_def #unit_name(val: #name) -> #quantity_type<#name, {#quantity_dimension}> {
                    #quantity_type::<#name, {#quantity_dimension}>(#value)
                }
            }
        }
    }

    fn vector_unit_constructor(
        &self,
        vector_type: &VectorType,
        unit: &Unit,
        quantity_dimension: &TokenStream,
    ) -> TokenStream {
        let Defs { quantity_type, .. } = &self;
        let Unit {
            name: unit_name,
            magnitude,
            ..
        } = unit;
        let VectorType {
            name,
            float_type,
            num_dims,
            ..
        } = &vector_type;
        let float_type = &float_type.name;
        let fn_args = match num_dims {
            2 => quote! { x: #float_type, y: #float_type },
            3 => quote! { x: #float_type, y: #float_type, z: #float_type },
            _ => unreachable!(),
        };
        let call_args = match num_dims {
            2 => quote! { x, y },
            3 => quote! { x, y, z },
            _ => unreachable!(),
        };
        let span = self.dimension_type.span();
        quote_spanned! {span =>
            impl #quantity_type<#name, {#quantity_dimension}> {
                pub fn #unit_name(#fn_args) -> #quantity_type<#name, {#quantity_dimension}> {
                    #quantity_type::<#name, {#quantity_dimension}>(#name::new(#call_args) * (#magnitude as #float_type))
                }
            }
        }
    }
}
