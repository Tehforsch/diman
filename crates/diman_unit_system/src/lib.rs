#![feature(proc_macro_diagnostic)]

mod codegen;
mod dimension_math;
mod parse;
mod resolve;
mod types;

use proc_macro2::TokenStream;
use syn::*;

#[proc_macro]
pub fn unit_system(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let defs = parse_macro_input!(item as types::UnresolvedTemplates);
    let expanded = defs.expand_templates();
    let resolved = expanded.resolve();
    let impls: TokenStream = resolved.code_gen();
    impls.into()
}
