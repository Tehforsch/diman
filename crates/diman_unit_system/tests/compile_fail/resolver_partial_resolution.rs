#![allow(incomplete_features)]
#![feature(generic_const_exprs, adt_const_params)]

use diman_unit_system::unit_system_internal;

unit_system_internal!(
    quantity_type Quantity;
    dimension_type Dimension;
    dimension Length;
    #[base(Length)]
    #[symbol(m)]
    unit meters;
    unit kilometers = 1000.0 * meters;
    unit millimeters = 0.001 * meters;
    unit undefined = undefined;
);

fn main() {
    use crate::units::meters;
    let l = 1.0 * meters;
}
