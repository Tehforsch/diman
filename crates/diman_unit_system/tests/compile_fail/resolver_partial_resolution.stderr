error: Unresolvable definition "undefined".
  --> tests/compile_fail/resolver_partial_resolution.rs:15:10
   |
15 |     unit undefined = undefined;
   |          ^^^^^^^^^
   |
   = help: Remove recursive definitions.
